src/main/scala/io/iohk/ethereum/network/discovery/codecs/RLPCodecs.scala

```diff
@@ -0,0 +1,181 @@
+import io.iohk.ethereum.rlp.RLPDecoder
+
+/** RLP codecs based on https://github.com/ethereum/devp2p/blob/master/discv4.md */
+object RLPCodecs {
+
+  implicit val inetAddressRLPCodec: RLPCodec[InetAddress] =
+    implicitly[RLPCodec[Array[Byte]]].xmap(InetAddress.getByAddress(_), _.getAddress)
+
+  implicit val bitVectorRLPCodec: RLPCodec[BitVector] =
+    implicitly[RLPCodec[Array[Byte]]].xmap(BitVector(_), _.toByteArray)
+
+  implicit val byteVectorRLPCodec: RLPCodec[ByteVector] =
+    implicitly[RLPCodec[Array[Byte]]].xmap(ByteVector(_), _.toArray)
+
+  implicit val hashRLPCodec: RLPCodec[Hash] =
+    implicitly[RLPCodec[BitVector]].xmap(Hash(_), identity)
+
+  implicit val publicKeyRLPCodec: RLPCodec[PublicKey] =
+    implicitly[RLPCodec[BitVector]].xmap(PublicKey(_), identity)
+
+  implicit val signatureRLPCodec: RLPCodec[Signature] =
+    implicitly[RLPCodec[BitVector]].xmap(Signature(_), identity)
+
+  implicit val nodeAddressRLPCodec: RLPCodec[Node.Address] =
+    deriveLabelledGenericRLPCodec
+
+  implicit val nodeRLPCodec: RLPCodec[Node] =
+    RLPCodec.instance[Node](
+      { case Node(id, address) =>
+        RLPEncoder.encode(address).asInstanceOf[RLPList] :+ id
+      },
+      {
+        case list @ RLPList(items @ _*) if items.length >= 4 =>
+          val address = RLPDecoder.decode[Node.Address](list)
+          val id = RLPDecoder.decode[PublicKey](items(3))
+          Node(id, address)
+      }
+    )
+
+  // https://github.com/ethereum/devp2p/blob/master/enr.md#rlp-encoding
+  // content = [seq, k, v, ...]
+  implicit val enrContentRLPCodec: RLPCodec[EthereumNodeRecord.Content] =
+    RLPCodec.instance(
+      { case EthereumNodeRecord.Content(seq, attrs) =>
+        val kvs = attrs
+          .foldRight(RLPList()) { case ((k, v), kvs) =>
+            k.toArray +: v.toArray +: kvs
+          }
+
+        seq +: kvs
+      },
+      { case RLPList(seq, kvs @ _*) =>
+        val attrs = kvs
+          .grouped(2)
+          .collect { case Seq(k, v) =>
+            rlp.decode[ByteVector](k) -> rlp.decode[ByteVector](v)
+          }
+          .toSeq
+
+        // TODO: Should have a constructor for key-value pairs.
+        import EthereumNodeRecord.byteOrdering
+
+        EthereumNodeRecord.Content(
+          seq,
+          SortedMap(attrs: _*)
+        )
+      }
+    )
+
+  // record = [signature, seq, k, v, ...]
+  implicit val enrRLPCodec: RLPCodec[EthereumNodeRecord] =
+    RLPCodec.instance(
+      { case EthereumNodeRecord(signature, content) =>
+        val contentList = RLPEncoder.encode(content).asInstanceOf[RLPList]
+        signature.toByteArray +: contentList
+      },
+      { case RLPList(signature, content @ _*) =>
+        EthereumNodeRecord(
+          rlp.decode[Signature](signature),
+          rlp.decode[EthereumNodeRecord.Content](RLPList(content: _*))
+        )
+      }
+    )
+
+  implicit val pingRLPCodec: RLPCodec[Payload.Ping] =
+    deriveLabelledGenericRLPCodec
```

Comment:
You mean that I should have classes that compare to hand-crafted `RLPList` structures? How would I know they are correct? I thought the test vectors I found in EIP8 are convincing enough that all these messages correspond to what they should be. 

I agree that it's somewhat convenient to see the mapping with your eyes, but in itself it doesn't convince at all that they correspond to the spec, there could still be a field missing or in the wrong order. It also looks like it could be easy to forget to include clause to capture the "rest" of the fields, like `RLPList(a, b, _*)` vs `RLPList(a, b)`, where the second version my work today but fail later if a new version adds more fields, or as EIP8 mandates that we have to accept random garbage at the end. [This commit](https://github.com/input-output-hk/mantis/pull/766/commits/cb228e347982033df4e86428d4bf6be87e7ea67a) had a version of hand-crafted codec, but with the trailing optional field it was a bit tricky.

I was hoping that the [ethereum tests](https://github.com/ethereum/tests) would have examples of everything but it doesn't seem to be the case.

Reply:
Another way I thought could be a bit less repetitive than what's in `.p2p.messages` is to use the [tuple codecs](https://github.com/input-output-hk/mantis/pull/766/files#diff-6d5ca06d8075843955d610fb56ee9ea34037bce67ce20beddae4f25cd2ff60c3R172) but I ended up not using them after all. With them the mapping would be something like this:

```scala
implicit val nodeAddressCodec = 
  implicitly[RLPCodec[(InetAddress, Int, Int)]].xmap(
   (Node.Address.apply _).tupled,
    Node.Address.unapply(_).get
  )
```

However I didn't try what kind of error message you'd get from that, while `deriveLabelledGenericRLPCodec` would tell you exactly which field failed to be decoded and why (it's better in that sense than the manual method relying on implicit conversions, any of which can fail at the same time).

Reply:
Added examples for what the encoded RLP looks like.

Reply:
sure i totally agree that test vectors are more reassuring than just looking at payloads (as this this tells nothing are those payloads transformed into actual bytes), but imo it is good to have the representation before your eyes in some cases. I think that simple test cases are good middle ground here.